src/Nethermind/Nethermind.Optimism/Rpc/OptimismEthRpcModule.cs

```diff
@@ -160,76 +157,86 @@ public override async Task<ResultWrapper<Hash256>> eth_sendRawTransaction(byte[]
         if (transaction is null)
         {
-            return ResultWrapper<OptimismTransactionForRpc?>.Success(null);
+            return ResultWrapper<TransactionForRpc?>.Success(null);
         }
 
         RecoverTxSenderIfNeeded(transaction);
-        OptimismTransactionForRpc transactionModel = new OptimismTransactionForRpc(transaction, blockHash: receipt?.BlockHash, receipt: receipt as OptimismTxReceipt);
+        var transactionModel = TransactionForRpc.FromTransaction(transaction, blockHash: receipt?.BlockHash);
+        if (transactionModel is DepositTransactionForRpc depositTx)
+        {
+            depositTx.DepositReceiptVersion = (receipt as OptimismTxReceipt)?.DepositReceiptVersion;
+        }
         if (_logger.IsTrace) _logger.Trace($"eth_getTransactionByHash request {transactionHash}, result: {transactionModel.Hash}");
-        return ResultWrapper<OptimismTransactionForRpc?>.Success(transactionModel);
+        return ResultWrapper<TransactionForRpc?>.Success(transactionModel);
     }
 
-    public new ResultWrapper<OptimismTransactionForRpc?> eth_getTransactionByBlockHashAndIndex(Hash256 blockHash, UInt256 positionIndex)
+    public override ResultWrapper<TransactionForRpc> eth_getTransactionByBlockHashAndIndex(Hash256 blockHash, UInt256 positionIndex)
     {
         SearchResult<Block> searchResult = _blockFinder.SearchForBlock(new BlockParameter(blockHash));
         if (searchResult.IsError || searchResult.Object is null)
         {
-            return GetFailureResult<OptimismTransactionForRpc?, Block>(searchResult, _ethSyncingInfo.SyncMode.HaveNotSyncedBodiesYet());
+            return GetFailureResult<TransactionForRpc, Block>(searchResult, _ethSyncingInfo.SyncMode.HaveNotSyncedBodiesYet());
         }
 
         Block block = searchResult.Object;
         if (positionIndex < 0 || positionIndex > block!.Transactions.Length - 1)
         {
-            return ResultWrapper<OptimismTransactionForRpc?>.Fail("Position Index is incorrect", ErrorCodes.InvalidParams);
+            return ResultWrapper<TransactionForRpc>.Fail("Position Index is incorrect", ErrorCodes.InvalidParams);
         }
 
-        OptimismTxReceipt[] receipts = _receiptFinder.Get(block).Cast<OptimismTxReceipt>().ToArray();
         Transaction transaction = block.Transactions[(int)positionIndex];
         RecoverTxSenderIfNeeded(transaction);
 
-        OptimismTransactionForRpc transactionModel = new OptimismTransactionForRpc(
-            transaction,
-            blockHash: block.Hash,
-            receipt: receipts.FirstOrDefault(r => r.TxHash == transaction.Hash));
+        var transactionModel = TransactionForRpc.FromTransaction(transaction, blockHash: block.Hash);
+        if (transactionModel is DepositTransactionForRpc depositTx)
+        {
+            OptimismTxReceipt? receipt = _receiptFinder
+                .Get(block)
+                .Cast<OptimismTxReceipt>()
```

Comment:
Are all receipts always of type `OptimismTxReceipt` or is it possible to have a `TxReceipt`? If the later is correct then this cast will fail at runtime.

cc. @flcl42 @jmederosalvarado